src/main/java/emissary/core/channels/BufferedChannelFactory.java

```diff
@@ -0,0 +1,132 @@
+         * 
+         * @param seekableByteChannelFactory to be cached.
+         * @param blockSize of the block of bytes.
+         */
+        public BufferedChannelFactoryImpl(final SeekableByteChannelFactory seekableByteChannelFactory,
+                final int blockSize) {
+            Validate.notNull(seekableByteChannelFactory, "Required: seekableByteChannelFactory not null!");
+            Validate.isTrue(blockSize > 0, "Required: blockSize > 0");
+
+            this.seekableByteChannelFactory = seekableByteChannelFactory;
+            this.blockSize = blockSize;
+        }
+
+        @Override
+        public SeekableByteChannel create() {
+            return new MetricsSeekableByteChannel(seekableByteChannelFactory.create(), blockSize);
+        }
+    }
+
+    /**
+     * SeekableByteChannel that caches the bytes of the passed in SeekableByteChannel
+     */
+    private static class MetricsSeekableByteChannel extends AbstractSeekableByteChannel {
+        /**
+         * The SeekableByteChannel to cache.
+         */
+        private final SeekableByteChannel seekableByteChannel;
+        /**
+         * The size of the block of bytes.
+         */
+        private final int blockSize;
+        /**
+         * The block of bytes.
+         */
+        private final ByteBuffer block;
+
+        /**
+         * The starting offset of the current block.
+         */
+        private long blockStart = -1;
+        /**
+         * The number of valid bytes in the current block.
+         */
+        private int blockValidBytes = -1;
+
+        /**
+         * Creates a SeekableByteChannel cache where there is a single block of bytes aligned on blockSize boundaries.
+         * 
+         * @param seekableByteChannel to be cached.
+         * @param blockSize of the block of bytes.
+         */
+        public MetricsSeekableByteChannel(final SeekableByteChannel seekableByteChannel, final int blockSize) {
+            this.seekableByteChannel = seekableByteChannel;
+            this.blockSize = blockSize;
+            this.block = ByteBuffer.allocate(blockSize);
+        }
+
+        @Override
+        protected void closeImpl() throws IOException {
+            seekableByteChannel.close();
+        }
+
+        @Override
+        protected int readImpl(final ByteBuffer byteBuffer, final int maxBytesToRead) throws IOException {
+            final long blockStartFromPosition = position() / blockSize * blockSize;
+
+            if (blockStartFromPosition != blockStart) {
+                block.position(0);
+                seekableByteChannel.position(blockStartFromPosition);
+
+                blockValidBytes = IOUtils.read(seekableByteChannel, block);
+                blockStart = blockStartFromPosition;
+            }
+
+            final int blockStartOffset = (int) (position() % blockSize);
+            final int bytesToReturn = Math.min(maxBytesToRead, blockValidBytes - blockStartOffset);
```

Comment:
Just to confirm, this method will only read up to `blockSize` bytes per call

Reply:
Yes.